pytorch/training/resnet.py

```diff
@@ -35,208 +17,121 @@ def forward(self, x): return x.view(x.size(0), -1)
-        self.bn2 = bn(planes)
-        self.conv3 = nn.Conv2d(planes, planes * 4, kernel_size=1, bias=False)
-        self.bn3 = bn(planes * 4, init_zero=True)
         self.relu = nn.ReLU(inplace=True)
-        self.downsample = downsample
         self.stride = stride
-
+        if self.stride == 1:
+            self.conv = nn.Conv2d(planes, planes, kernel_size=(1,3), stride=stride,
+                                   padding=(0,1), bias=False)
+            self.bn = nn.BatchNorm2d(planes)
+            self.conv1 = nn.Conv2d(planes, planes, kernel_size=(3,1), stride=stride,
+                                   padding=(1,0), bias=False)
+            self.bn1 = nn.BatchNorm2d(planes)
+        else:
+            self.conv = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
+                                   padding=1, bias=False)
+            self.bn = nn.BatchNorm2d(planes)
     def forward(self, x):
-        residual = x
-        if self.downsample is not None: residual = self.downsample(x)
-
-        out = self.conv1(x)
-        out = self.bn1(out)
+        out = self.conv(x)
+        out = self.bn(out)
         out = self.relu(out)
-
-        out = self.conv2(out)
-        out = self.bn2(out)
-        out = self.relu(out)
-
-        out = self.conv3(out)
-        out = self.bn3(out)
-
-        out += residual
-        out = self.relu(out)
-
+        if self.stride == 1:
+            out = self.conv1(out)
+            out = self.bn1(out)
+            out = self.relu(out)
         return out
+            
 
-class Bottleneck(nn.Module):
+class BottleneckFactorized(nn.Module):
     expansion = 4
 
     def __init__(self, inplanes, planes, stride=1, downsample=None):
-        super().__init__()
+        super(BottleneckFactorized, self).__init__()
         self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
-        self.bn1 = bn(planes)
-        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
-                               padding=1, bias=False)
-        self.bn2 = bn(planes)
-        self.conv3 = nn.Conv2d(planes, planes * 4, kernel_size=1, bias=False)
-        self.bn3 = bn(planes * 4)
+        self.bn1 = nn.BatchNorm2d(planes)
+        self.conv2 = FConv3(planes, stride)
+        self.conv3 = nn.Conv2d(planes, planes * self.expansion, kernel_size=1, bias=False)
+        self.bn3 = nn.BatchNorm2d(planes * self.expansion)
         self.relu = nn.ReLU(inplace=True)
         self.downsample = downsample
         self.stride = stride
 
     def forward(self, x):
         residual = x
-        if self.downsample is not None: residual = self.downsample(x)
 
         out = self.conv1(x)
         out = self.bn1(out)
         out = self.relu(out)
-
         out = self.conv2(out)
-        out = self.bn2(out)
-        out = self.relu(out)
 
         out = self.conv3(out)
         out = self.bn3(out)
 
+        if self.downsample is not None:
+            residual = self.downsample(x)
```

Comment:
This is basically the exact code from torchvision: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py

I think downsample is just a conv layer (passed in) with stride of 2. I think they pass it in because not all the blocks need it